### PR TITLE
integrate with ALD IFE

### PR DESCRIPTION
**Pull request by pnowosie**
EDIT: pdobacz: for full PR description see below. The opening comment is Paweł's original draft PR's comment :point_down: 

Beginning of the journey to integrate ALD's inflight exits

Changes between old IFE events and ALD's ones -> check this [gist](https://gist.github.com/pnowosie/18c8cdf10fffa123f4aaeaa0ce203689)

I managed:
- [root_chain.ex](https://github.com/omisego/elixir-omg/pull/1032/commits/698b1d423397de876eb66a829b160b433b09c0f2) - which fetches lists of new events for both `inputs/outputs` and concatenate them
- started to change [inflight_exit_info](https://github.com/omisego/elixir-omg/pull/1032/commits/afd038e2e8150b8f54e0dfce68d89287692ead6a) where `output_index` is expressed as a 
tuple: `{:input | :output, oindex}`

### CODE REVIEW

File: apps/omg_eth/lib/omg_eth/root_chain.ex [hunk 1]
```diff
@@ -144,16 +149,6 @@ defmodule OMG.Eth.RootChain do
          do: {:ok, Enum.map(all_sorted_logs, &decode_deposit/1)}
   end
 
-  @spec get_piggybacks(non_neg_integer, non_neg_integer, optional_addr_t) ::
```
Comment (pnowosie) [inline apps/omg_eth/lib/omg_eth/root_chain.ex]: Moved below :point_down:

File: apps/omg_watcher/lib/omg_watcher/exit_processor/in_flight_exit_info.ex [hunk 2]
```diff
@@ -80,7 +71,7 @@ defmodule OMG.Watcher.ExitProcessor.InFlightExitInfo do
           oldest_competitor: Utxo.Position.t() | nil,
           eth_height: pos_integer(),
           exit_map: %{
-            non_neg_integer() => %{
+            {:input | :output, non_neg_integer()} => %{
```
Comment (pnowosie) [inline apps/omg_watcher/lib/omg_watcher/exit_processor/in_flight_exit_info.ex]: :exclamation: Idea to change `output index` from integer in `0..7` to tuple: `{:input | :output, integer in 0..4}`

**Comment by pdobacz** [comment]
#987 

## Overview

This allows `elixir-omg` to integrate with the ALD contracts in terms of in-flight exit operations.

The change is roughly split into two parts - the integration and some refactors that became necessary.

## Changes

Main part is `WIP: unfinished changes in handling piggybacks` to `feat: support using information about input txs for competitor call data`, where the integration is delivered. WIthin this there are 3 commits related to the tentative change where `input_tx` and position are required for input challenges: 
- `feat: support holding onto the information about input txs (tentative)`
- `feat: support using information about input txs for competitor challenge`
- `feat: support using information about input txs for competitor call data`
**Those 3 commits might end up reverted**, in case we opt for undoing this requirement on the contract's end.

Then the refactors:
 - `refactor: tidy Eth.RootChain - the more complicated log decoding routine` splits some `OMG.Eth` code into pieces
 - `fix: tidy the behavior of finalizing piggybacked outputs` - this is a more involved one. This changes the approach to finalize piggybacked outputs in the watcher. Instead of feeding the watcher with raw piggybacking events, that were artificially recreated, we're now leveraging the tx inclusion information that is anyway visible to the `ExitProcessor`. Thanks to this, we can do the finalization based on the utxo positions much simpler. This wasn't strictly a refactor, b/c it actually fixed one issue noted in #994. See changes to the integration tests for details. `refactor: don't treat a non-modified piece of data as reduce result` through `refactor: tidy the processor's interim data structs after refactors` tidy this change up.
 - `refactor: initalize EthereumEventListener.Groomer for event processing` introduces an interim layer for pre-processing of contract event data. This is used to make `OMG.Eth` lighter and also put some load off of core modules in the future
 - `refactor: remove :last_deposit_child_blknum being stored, was redundant` as title, this was a redundant runtime check and hard to maintain
 - and some more minor fixes/refactors on top of that

The swagger definitions were updated.



## Testing

`mix test && mix test --only integration`

**Comment by pdobacz** [comment]
TODO:
 - [ ] rebase on `plasma-contracts/master` after both https://github.com/omisego/plasma-contracts/issues/344 and https://github.com/omisego/plasma-contracts/issues/342 land (this is causing the builds to fail)

**Comment by pdobacz** [comment]
eh, need to close & reopen to ask for review

**Comment by pdobacz** [comment]
moved to #1046

**Pull request closed without merging**